### PR TITLE
fix: sync ovn-ic static route to policy

### LINKED ISSUE #1589: lr-policy breaks ovn-ic interconnection
# Expected Behavior

# Actual Behavior

# Steps to Reproduce the Problem

1.
2.
3.

# Additional Info

- Kubernetes version:

  **Output of `kubectl version`:**

  ```
  (paste your output here)
  ```

- kube-ovn version:
  
  ```
  (paste your output here)
  ```

- operation-system/kernel version:
  
  **Output of `awk -F '=' '/PRETTY_NAME/ { print $2 }' /etc/os-release`:**
  **Output of `uname -r`:**
  ```
  (paste your output here)
  ```

### PR DESCRIPTION
**Pull request by lut777**
#### What type of this PR
Examples of user facing changes:
- Bug fixes

#### Which issue(s) this PR fixes:
Fixes #1589

interval time of each sync is set to be 5 seconds now. Considering the small number of sync routes, this sync time should not need to be modified.

### CODE REVIEW

File: pkg/controller/ovn-ic.go [hunk 1]
```diff
@@ -361,3 +365,75 @@ func genHostAddress(host string, port string) (hostaddress string) {
...
+	lrRouteList, err := c.ovnClient.GetLogicalRouterRouteByOptions(util.OVNIC_Key, util.OVNIC_Value)
+	if err != nil {
+		klog.V(5).Infof("failed to list lr ovn-ic route %v", err)
+		return
+	}
+	if len(lrRouteList) == 0 {
+		klog.V(5).Info(" lr ovn-ic route does not exist")
+		lrPolicyList, err := c.ovnClient.GetLogicalRouterPoliciesByOptions(util.OVNIC_Key, util.OVNIC_Value)
+		if err != nil {
+			klog.Warning("failed to list ovn-ic lr policy ", err)
```
Comment (oilbeater) [inline pkg/controller/ovn-ic.go:385]: klog.Errorf

File: pkg/controller/ovn-ic.go [hunk 2]
```diff
@@ -361,3 +365,75 @@ func genHostAddress(host string, port string) (hostaddress string) {
...
+			return
+		}
+		for _, lrPolicy := range lrPolicyList {
+			err := c.ovnClient.DeleteRouterPolicy(lr, lrPolicy.UUID)
```
Comment (oilbeater) [inline pkg/controller/ovn-ic.go:389]: if err := xxx; err != nil {}

File: pkg/controller/ovn-ic.go [hunk 3]
```diff
@@ -361,3 +365,75 @@ func genHostAddress(host string, port string) (hostaddress string) {
...
+			if err != nil {
+				klog.Warningf("deleting router policy failed %v", err)
```
Comment (oilbeater) [inline pkg/controller/ovn-ic.go:391]: klog.Errorf

File: pkg/ovs/ovn-nb-logical_router_policy.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/ovn-org/libovsdb/model"
+	"github.com/ovn-org/libovsdb/ovsdb"
+)
+
+func (c OvnClient) AddRouterPolicy(lr *ovnnb.LogicalRouter, matchfield string, action ovnnb.LogicalRouterPolicyAction,
+	opts map[string]string, extIDs map[string]string, priority int) error {
+	lrPolicy := &ovnnb.LogicalRouterPolicy{
+		Action:      action,
+		Match:       matchfield,
+		Options:     opts,
```
Comment (oilbeater) [inline pkg/ovs/ovn-nb-logical_router_policy.go:16]: should set external_ids not options

File: pkg/controller/ovn-ic.go [hunk 5]
```diff
@@ -361,3 +365,75 @@ func genHostAddress(host string, port string) (hostaddress string) {
...
+			}
+		}
+		return
+	}
+
+	policyMap := map[string]string{}
+	lrPolicyList, err := c.ovnClient.GetLogicalRouterPoliciesByOptions(util.OVNIC_Key, util.OVNIC_Value)
+	if err != nil {
+		klog.Warning("failed to list ovn-ic lr policy ", err)
```
Comment (oilbeater) [inline pkg/controller/ovn-ic.go:400]: all the warning should be replaced with Errorf

File: pkg/controller/ovn-ic.go [hunk 6]
```diff
@@ -361,3 +365,75 @@ func genHostAddress(host string, port string) (hostaddress string) {
...
+			err := c.ovnClient.AddRouterPolicy(lr, matchFiled, ovnnb.LogicalRouterPolicyActionAllow,
+				map[string]string{util.OVNIC_Key: util.OVNIC_Value, "vendor": util.CniTypeName},
+				map[string]string{}, util.OvnICPolicyPriority)
+			if err != nil {
+				klog.Warningf("adding router policy failed %v", err)
+			}
+		}
+	}
+	for _, uuid := range policyMap {
+		err := c.ovnClient.DeleteRouterPolicy(lr, uuid)
```
Comment (oilbeater) [inline pkg/controller/ovn-ic.go:425]: if err := xxx; err != nil {}

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: pkg/util/const.go [hunk 7]
```diff
@@ -169,4 +170,10 @@ const (
 	POD_IP             = "POD_IP"
 	ContentType        = "application/vnd.kubernetes.protobuf"
 	AcceptContentTypes = "application/vnd.kubernetes.protobuf,application/json"
+
+	OVNIC_Key   = "origin"
+	OVNIC_Value = "connected"
```
[the remark this answers is not in the document]
Comment (lut777) [reply pkg/util/const.go]: done

File: pkg/controller/ovn-ic.go [hunk 8]
```diff
@@ -361,3 +365,75 @@ func genHostAddress(host string, port string) (hostaddress string) {
 	}
 	return hostaddress
 }
+
+func (c *Controller) SynRouteToPolicy() {
+
+	lr, err := c.ovnClient.GetLogicalRouter(util.DefaultVpc, false)
+	if err != nil {
+		klog.V(5).Infof("logical router does not exist %v at %v", err, time.Now())
```
[the remark this answers is not in the document]
Comment (lut777) [reply pkg/controller/ovn-ic.go:373]: done

File: pkg/controller/ovn-ic.go (hunk 1 above)
Comment (lut777) [reply pkg/controller/ovn-ic.go]: done

File: pkg/controller/ovn-ic.go (hunk 2 above)
Comment (lut777) [reply pkg/controller/ovn-ic.go]: done

File: pkg/controller/ovn-ic.go (hunk 3 above)
Comment (lut777) [reply pkg/controller/ovn-ic.go]: done

File: pkg/controller/ovn-ic.go (hunk 5 above)
Comment (lut777) [reply pkg/controller/ovn-ic.go]: done

File: pkg/controller/ovn-ic.go (hunk 6 above)
Comment (lut777) [reply pkg/controller/ovn-ic.go]: done

File: pkg/ovs/ovn-nb-logical_router_policy.go (hunk 4 above)
Comment (lut777) [reply pkg/ovs/ovn-nb-logical_router_policy.go:16]: external_id is listed here for a complete installation. Actually it is not used.

**Pull request merged**